cmd/disk-cache.go

```diff
@@ -0,0 +1,680 @@
+	}
+	objInfo, err := GetObjectInfo(bucket, object)
+	_, backendDown := errorCause(err).(BackendDown)
+	if err != nil && !backendDown {
+		if _, ok := errorCause(err).(ObjectNotFound); ok {
+			// Delete the cached entry if backend object was deleted.
+			c.dcache.Delete(bucket, object)
+		}
+		return err
+	}
+	r, cachedObjInfo, anon, err := c.dcache.Get(bucket, object)
+	if err == nil {
+		defer r.Close()
+		if backendDown {
+			if anonReq {
+				if anon {
+					_, err = io.Copy(writer, io.NewSectionReader(r, startOffset, length))
+					return err
+				}
+				return BackendDown{}
+			}
+			// If the backend is down, serve the request from cache.
+			_, err = io.Copy(writer, io.NewSectionReader(r, startOffset, length))
+			return err
+		}
+		if cachedObjInfo.ETag == objInfo.ETag {
+			_, err = io.Copy(writer, io.NewSectionReader(r, startOffset, length))
+			return err
+		}
+		c.dcache.Delete(bucket, object)
+	}
+	if startOffset != 0 || length != objInfo.Size {
+		// We don't cache partial objects.
+		return GetObjectFn(bucket, object, startOffset, length, writer)
+	}
+	cachedObj, err := c.dcache.Put(length)
+	if err == errDiskFull {
+		return GetObjectFn(bucket, object, 0, objInfo.Size, writer)
+	}
+	if err != nil {
+		return err
+	}
+	err = GetObjectFn(bucket, object, 0, objInfo.Size, io.MultiWriter(writer, cachedObj))
+	if err != nil {
+		c.dcache.NoCommit(cachedObj)
+		return err
+	}
+	objInfo, err = GetObjectInfoFn(bucket, object)
+	if err != nil {
+		c.dcache.NoCommit(cachedObj)
+		return err
+	}
+	// FIXME: race-condition: what if the server object got replaced between GetObjectFn and GetObjectInfoFn
+	return c.dcache.Commit(cachedObj, objInfo, true)
+}
+
+// Uses cached-object to serve the request. If object is not cached it serves the request from the backend and also
+// stores it in the cache for serving subsequent requests.
+func (c cacheObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
+	return c.getObject(bucket, object, startOffset, length, writer, false)
+}
+
+// Anonymous version of cacheObjects.GetObject
+func (c cacheObjects) AnonGetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
+	return c.getObject(bucket, object, startOffset, length, writer, true)
+}
+
+// Common function for GetObjectInfo and AnonGetObjectInfo
+func (c cacheObjects) getObjectInfo(bucket, object string, anonReq bool) (ObjectInfo, error) {
+	getObjectInfoFn := c.GetObjectInfoFn
+	if anonReq {
+		getObjectInfoFn = c.AnonGetObjectInfoFn
+	}
+	objInfo, err := getObjectInfoFn(bucket, object)
+	if err != nil {
+		if _, ok := errorCause(err).(BackendDown); !ok {
+			if _, ok = errorCause(err).(ObjectNotFound); ok {
+				// Delte the cached entry if backend object was deleted.
+				c.dcache.Delete(bucket, object)
+			}
+			return ObjectInfo{}, err
+		}
+		cachedObjInfo, anon, infoErr := c.dcache.GetObjectInfo(bucket, object)
+		if infoErr == nil {
+			if !anonReq {
```

Comment:
What is the difference between anon and anonReq?  perhaps we can choose a better variable name to differentiate their purpose.